src/lib/dump_dir.c

```diff
@@ -2425,110 +2426,130 @@ int dd_create_archive(struct dump_dir *dd, const char *archive_name,
     }
 
     /* Write data to the tarball */
     dd_init_next_file(dd);
     char *short_name, *full_name;
+    entry = archive_entry_new();
+    if (entry == NULL)
+    {
+        result = -1;
+        log_warning(_("Failed to allocate and initialize archive entry object"));
+        return result;
+
+    }
+
     while (dd_get_next_file(dd, &short_name, &full_name))
     {
         if (!(exclude_elements && libreport_is_in_string_list(short_name, exclude_elements)))
         {
-           if (tar_append_file(tar, full_name, short_name))
-               result = -errno;
+            stat(full_name, &st);
+            archive_entry_set_pathname(entry, short_name);
+            archive_entry_copy_stat(entry, &st);
+            archive_entry_set_filetype(entry, AE_IFREG);
+            archive_entry_set_perm(entry, 0644);
+
+retry:
+            r = archive_write_header(a, entry);
+            switch (r)
+            {
+                case ARCHIVE_WARN:
+                    log_warning(_("%s"), archive_error_string(a));
+                case ARCHIVE_OK:
+                    break;
+                case ARCHIVE_RETRY:
+                    if (++tries < max_tries)
+                    {
+                        log_warning(_("Failed to write to archive, retrying..."));
+                        goto retry;
+                    }
+                case ARCHIVE_FATAL:
+                    log_warning(_("Failed to write to archive: %s"),
+                            (tries == max_tries ?  "too many attempts" : archive_error_string(a)));
+                    result = (tries == max_tries ? -1 : -archive_errno(a));
+                    goto finito;
+            }
+
+            archive_entry_clear(entry);
+            fd = open(full_name, O_RDONLY);
+            if (fd == -1)
+            {
+                result = -errno;
+                log_warning(_("Failed to add file to archive: %s"), full_name);
+                goto finito;
+            }
+            len = read(fd, buff, sizeof (buff));
+            while (len > 0)
+            {
+                archive_write_data(a, buff, len);
+                len = read(fd, buff, sizeof (buff));
+            }
+            close(fd);
         }
 
         free(short_name);
         free(full_name);
 
-        if (result != 0)
-            goto finito;
-    }
-
-    /* Close tar writer... */
-    if (tar_append_eof(tar) != 0)
-    {
-        result = -errno;
-        log_warning(_("Failed to finalize TAR archive"));
-        goto finito;
     }
 
 finito:
-    signal(SIGPIPE, old_handler);
-
-    if (tar != NULL && tar_close(tar) != 0)
-    {
-        result = -errno;
-        log_warning(_("Failed to close TAR writer"));
-    }
 
-    /* ...and check that gzip child finished successfully */
-    int status;
-    libreport_safe_waitpid(child, &status, 0);
-    if (status != 0)
+    archive_entry_free(entry);
+    archive_write_close(a);
+    archive_write_free(a);
+    r = chmod(archive_name, 0600);
```

Comment:
I'm thinking, for the sake of security, we should call `open()` with these permissions (as in the previous version) and then `archive_write_open_fd()`. Otherwise the file might be readable before the archive is finalised.

Reply:
> I'm thinking, for the sake of security, we should call `open()` with these permissions (as in the previous version) and then `archive_write_open_fd()`. Otherwise the file might be readable before the archive is finalised.  
  
Good point.  
  
I'm not sure about the `close()`.